Signal/src/ViewControllers/OWSAddToContactViewController.m

```diff
@@ -0,0 +1,250 @@
+        // Saving normally returns you to the "Show Contact" view
+        // which we're not interested in, so we skip it here. There is
+        // an unfortunate blip of the "Show Contact" view on slower devices.
+        DDLogDebug(@"%@ completed editing contact.", self.tag);
+        [self dismissViewControllerAnimated:NO
+                                 completion:^{
+                                     [self.navigationController popViewControllerAnimated:YES];
+                                 }];
+    } else {
+        DDLogDebug(@"%@ canceled editing contact.", self.tag);
+        [self dismissViewControllerAnimated:YES
+                                 completion:^{
+                                     [self.navigationController popViewControllerAnimated:YES];
+                                 }];
+    }
+}
+
+#pragma mark - ContactsViewHelperDelegate
+
+- (void)contactsViewHelperDidUpdateContacts
+{
+    [self updateTableContents];
+}
+
+#pragma mark - View Lifecycle
+
+- (void)viewDidLoad
+{
+    [super viewDidLoad];
+
+    self.title = NSLocalizedString(@"CONVERSATION_SETTINGS_ADD_TO_EXISTING_CONTACT",
+        @"Label for 'new contact' button in conversation settings view.");
+
+    [self updateTableContents];
+}
+
+- (nullable NSString *)displayNameForContact:(Contact *)contact
+{
+    OWSAssert(contact);
+
+    if (contact.fullName.length > 0) {
+        return contact.fullName;
+    }
+
+    for (NSString *email in contact.emails) {
+        if (email.length > 0) {
+            return email;
+        }
+    }
+    for (NSString *phoneNumber in contact.userTextPhoneNumbers) {
+        if (phoneNumber.length > 0) {
+            return phoneNumber;
+        }
+    }
+
+    return nil;
+}
+
+- (void)updateTableContents
+{
+    OWSTableContents *contents = [OWSTableContents new];
+    contents.title = NSLocalizedString(@"CONVERSATION_SETTINGS", @"title for conversation settings screen");
+
+    __weak OWSAddToContactViewController *weakSelf = self;
+
+    OWSTableSection *section = [OWSTableSection new];
+    section.headerTitle = NSLocalizedString(
+        @"EDIT_GROUP_CONTACTS_SECTION_TITLE", @"a title for the contacts section of the 'new/update group' view.");
+
+    NSMutableArray<NamedContact *> *namedContacts = [NSMutableArray new];
+    for (Contact *contact in self.contactsViewHelper.contactsManager.allContacts) {
+        OWSAssert(contact.cnContact);
+
+        NSString *_Nullable displayName = [self displayNameForContact:contact];
+        if (displayName.length < 1) {
+            continue;
+        }
+
+        NamedContact *namedContact = [NamedContact new];
+        namedContact.contact = contact;
+        namedContact.displayName = displayName;
+        [namedContacts addObject:namedContact];
+    }
+
+    [namedContacts sortUsingComparator:^NSComparisonResult(NamedContact *_Nonnull left, NamedContact *_Nonnull right) {
+        return [left.displayName caseInsensitiveCompare:right.displayName];
```

Comment:
You're right.  I've taken out the logic to do local sorting here.  

> ContactsManager.allContacts is already sorted in this order, but it don't think it accounts for sorting unnamed contacts by email/phone number.

The sorting done in `ContactsFrameworkContactStoreAdaptee` works pretty well with "un-named" contacts.